oak-store-document/src/main/java/org/apache/jackrabbit/oak/plugins/document/util/ThrottlingDocumentStoreWrapper.java

```diff
@@ -0,0 +1,216 @@
+        performThrottling();
+        return store.create(collection, updateOps);
+    }
+
+    @Override
+    public <T extends Document> T createOrUpdate(final Collection<T> collection, final UpdateOp update) {
+        performThrottling();
+        return store.createOrUpdate(collection, update);
+    }
+
+    @Override
+    public <T extends Document> List<T> createOrUpdate(final Collection<T> collection, final List<UpdateOp> updateOps) {
+        performThrottling();
+        return store.createOrUpdate(collection, updateOps);
+    }
+
+    @Override
+    public <T extends Document> T findAndUpdate(final Collection<T> collection, final UpdateOp update) {
+        performThrottling();
+        return store.findAndUpdate(collection, update);
+    }
+
+    @Override
+    public CacheInvalidationStats invalidateCache() {
+        return store.invalidateCache();
+    }
+    
+    @Override
+    public CacheInvalidationStats invalidateCache(Iterable<String> keys) {
+        return store.invalidateCache(keys);
+    }
+
+    @Override
+    public <T extends Document> void invalidateCache(Collection<T> collection, String key) {
+        store.invalidateCache(collection, key);
+    }
+
+    @Override
+    public void dispose() {
+        store.dispose();
+    }
+
+    @Override
+    public <T extends Document> T getIfCached(final Collection<T> collection, final String key) {
+        return store.getIfCached(collection, key);
+    }
+
+    @Override
+    public void setReadWriteMode(String readWriteMode) {
+        store.setReadWriteMode(readWriteMode);
+    }
+
+    @Override
+    public Iterable<CacheStats> getCacheStats() {
+        return store.getCacheStats();
+    }
+
+    @Override
+    public Map<String, String> getMetadata() {
+        return store.getMetadata();
+    }
+
+    @NotNull
+    @Override
+    public Map<String, String> getStats() {
+        return store.getStats();
+    }
+
+    @Override
+    public long determineServerTimeDifferenceMillis() {
+        return store.determineServerTimeDifferenceMillis();
+    }
+
+    /**
+     * Return the size limit for node name based on the document store implementation
+     *
+     * @return node name size limit
+     */
+    @Override
+    public int getNodeNameLimit() {
+        return store.getNodeNameLimit();
+    }
+
+    /**
+     * Return the {@link Throttler} for the underlying store
+     * Default is no throttling
+     *
+     * @return throttler for document store
+     */
+    @Override
+    public Throttler throttler() {
+        return store.throttler();
+    }
+
+    // helper methods
+
+    private void performThrottling() {
+
+        final Throttler throttler = throttler();
+        long throttleTime = throttler.throttlingTime();
+
+        if (throttleTime == 0) {
+            return; // no throttling
+        }
+
+        try {
+            LOG.info("Throttling the system for {} ms", throttleTime);
```

Comment:
I agree that we should log at INFO if throttling is happening. I'm wondering though how noisy this one could become. Could we reduce the frequency of how much we log at INFO, for example only every 10 sec (without causing too much synchronization between the different threads calling `performThrottle`..)?

Reply:
on 2nd thought, perhaps this even qualifies for a WARN?